src/mon/Monitor.cc

```diff
@@ -247,6 +248,7 @@ Monitor::Monitor(CephContext* cct_, string nm, MonitorDBStore *s,
   paxos_service[PAXOS_HEALTH].reset(new HealthMonitor(*this, *paxos, "health"));
   paxos_service[PAXOS_CONFIG].reset(new ConfigMonitor(*this, *paxos, "config"));
   paxos_service[PAXOS_KV].reset(new KVMonitor(*this, *paxos, "kv"));
+  paxos_service[PAXOS_NVMEGW].reset(new NVMeofGwMon(*this, *paxos, "nvmeofgw"));
```

Comment:
I would argue for having run-time optionality for registering `NVMeofGwMon`. Without it the service affects everyone, even those who don't make any use of NVMe-oF.

Let's introduce a configurable like `mon_service_nvmeofgwmon` (better name wanted!) with the default being `false`. In the QA suite we toggle it to bring coverage for the service.

`paxos_service` is:

```cpp
std::array<std::unique_ptr<PaxosService>, PAXOS_NUM> paxos_service;
```

so call sites would require `null` checks which shouldn't be a big deal as there are few of them.

Reply:
@rzarzynski  I like the idea and we should implement it in a separate PR as it is a behavior change.
At the moment all services are registered and changing that must be tested thoroughly. 
This PR is large and complex as is.
We should implement it for the MDS service for clusters without CephFS enabled.

Reply:
>  This PR is large and complex as is.

Yes, and its complexity is precisely the reason behind turning the mon service off by default.

>  We should implement it for the MDS service for clusters without CephFS enabled.

We could but the question is about benefits. For MDS, after all the years, they would be mostly in terms of footprint optimization; for `NVMeofGwMon` they are about risk management.

Reply:
@rzarzynski thank you for your help! Let me know if the [diff](https://github.com/ceph/ceph/pull/54671/commits/3bc0613b7047b77a67d79ceaa7a31b68be4d62af) looks OK.

Reply:
I don't think a configurable makes sense for this. It should be a setting in the monmap which enables or disables a service like the mgr. We can't have disagreement between monitors about which PaxosServices to operate.

Reply:
@rzarzynski @caroav @oritwas 

I have some concerns about the proposed [diff](https://github.com/ceph/ceph/pull/54671/commits/3bc0613b7047b77a67d79ceaa7a31b68be4d62af). Introducing null pointers into the Monitor core, where we’ve traditionally assumed pointers are always valid, makes me move uneasy 😳. The extensive changes to the monitor core with the addition of null checks seem risky.

I believe this approach could be handled in a separate PR, considering the existing Ceph monitor infrastructure supports dynamic services. The fact that this PR code passed QA tests in the 7.1 downstream version gives us additional confidence. Please let me know your thoughts on this.

Reply:
@rzarzynski @athanatos thank you for your help 🖖.  Following our discussion yesterday, please review the compile-time flag for the NVMeOF gateway monitor code, which is [disabled by default](https://github.com/baum/ceph/blob/007dfb45a6942a1c2ff04b4d18f3d3810eda3afe/src/mon/Monitor.h#L64-L66).

See the [diff](https://github.com/ceph/ceph/pull/54671/commits/007dfb45a6942a1c2ff04b4d18f3d3810eda3afe). I would appreciate your feedback.

Reply:
I'm happy with the `ifdefs`.